### PR TITLE
Input Shape Check for kNN Predict

### PR DESCRIPTION
**Pull request by owerbat**

### CODE REVIEW

File: daal4py/sklearn/neighbors/knn.py [hunk 1]
```diff
@@ -209,3 +213,60 @@ def predict(self, X):
         else:
             logging.info("sklearn.neighbors.KNeighborsClassifier.predict: " + method_uses_sklearn)
             return super(KNeighborsClassifier, self).predict(X)
+
+    def predict_proba(self, X):
```
Comment (Alexsandruss) [inline daal4py/sklearn/neighbors/knn.py:217]: Why this method is needed? It's optimized through self.kneighbors

File: daal4py/sklearn/neighbors/_base.py [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+# ******************************************************************************/
+
+
+import numpy as np
+
+
+def _get_weights(dist, weights):
```
Comment (Alexsandruss) [inline daal4py/sklearn/neighbors/_base.py:22]: Why this function is needed if predict_proba will not be patched?

File: daal4py/sklearn/neighbors/knn.py [hunk 3]
```diff
@@ -161,6 +162,9 @@ class KNeighborsClassifier(BaseKNeighborsClassifier, KNeighborsMixin):
     def predict(self, X):
         X = check_array(X, accept_sparse='csr')
 
+        if hasattr(self, 'n_features_in_') and X.shape[1] != self.n_features_in_:
```
Comment (Alexsandruss) [inline daal4py/sklearn/neighbors/knn.py:165]: Does 'n_features_in_' always exist and have correct value in this class?

File: daal4py/sklearn/neighbors/knn.py [hunk 4]
```diff
@@ -209,3 +213,60 @@ def predict(self, X):
...
+        -------
+        p : ndarray of shape (n_queries, n_classes), or a list of n_outputs
+            of such arrays if n_outputs > 1.
+            The class probabilities of the input samples. Classes are ordered
+            by lexicographic order.
+        """
+        X = check_array(X, accept_sparse='csr')
+
+        if hasattr(self, 'n_features_in_') and X.shape[1] != self.n_features_in_:
+            raise ValueError('Input data shape {} is inconsistent with the trained model'.format(X.shape))
```
Comment (Alexsandruss) [inline daal4py/sklearn/neighbors/knn.py:236]: Try to move to kneighbors method

Comment (owerbat) [reply daal4py/sklearn/neighbors/knn.py]: Thanks, it should work

File: daal4py/sklearn/neighbors/knn.py (hunk 3 above)
Comment (owerbat) [reply daal4py/sklearn/neighbors/knn.py]: As I saw in tests it doesn't always exist. That's why I'm using `getattr`

File: daal4py/sklearn/neighbors/knn.py (hunk 1 above)
Comment (owerbat) [reply daal4py/sklearn/neighbors/knn.py]: Removed

File: daal4py/sklearn/neighbors/_base.py (hunk 2 above)
Comment (owerbat) [reply daal4py/sklearn/neighbors/_base.py]: Removed

File: daal4py/sklearn/neighbors/knn.py [hunk 5]
```diff
@@ -32,6 +32,11 @@
 
 class KNeighborsMixin(BaseKNeighborsMixin):
     def kneighbors(self, X=None, n_neighbors=None, return_distance=True):
+        n_features = getattr(self, 'n_features_in_', None)
+        shape = getattr(X, 'shape', None)
+        if n_features and shape and shape[1] != n_features:
```
Comment (Alexsandruss) [inline daal4py/sklearn/neighbors/knn.py:37]: len(shape) might be 1 if 1d array is passed and error will be raised while trying to get shape[1]

File: daal4py/sklearn/neighbors/knn.py [hunk 6]
```diff
@@ -161,6 +166,9 @@ class KNeighborsClassifier(BaseKNeighborsClassifier, KNeighborsMixin):
     def predict(self, X):
         X = check_array(X, accept_sparse='csr')
 
+        # if hasattr(self, 'n_features_in_') and X.shape[1] != self.n_features_in_:
+        #     raise ValueError('Input data shape {} is inconsistent with the trained model'.format(X.shape))
+
```
Comment (Alexsandruss) [inline daal4py/sklearn/neighbors/knn.py:171]: Delete or uncomment

File: daal4py/sklearn/neighbors/knn.py (hunk 5 above)
Comment (owerbat) [reply daal4py/sklearn/neighbors/knn.py]: Additional check was added

**Comment by Alexsandruss** [comment]
predict method of classifier should have same check.

**Review by Alexsandruss — APPROVED** [review]

**Pull request merged**